### PR TITLE
[MRG+1] Fix a number of Deprecated/Invalid escape sequences

### PR DESCRIPTION
**Pull request by Carreau**
And turn python 3.6 DeprecationWarnings into error during tests.

Python 3.6 deprecate invalid escape sequences, as they are (most of the
time) mistakes from the author. This introduce extra `\` when necessary,
or make strings `raw`.

See #7924

### CODE REVIEW

File: lib/matplotlib/backends/backend_ps.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from __future__ import (absolute_import, division, print_function,
                         unicode_literals)
 
+
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py]: Unnecessary change.

File: lib/matplotlib/backends/backend_ps.py [hunk 2]
```diff
@@ -1543,7 +1544,6 @@ def gs_distill(tmpfile, eps=False, ptype='letter', bbox=None, rotated=False):
         else:
             pstoeps(tmpfile)
 
-
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py]: Should not be changed.

File: lib/matplotlib/font_manager.py [hunk 3]
```diff
@@ -184,7 +184,7 @@ def win32FontDirectory():
     Return the user-specified font directory for Win32.  This is
     looked up from the registry key::
 
-      \\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\Shell Folders\Fonts
+      \\HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\Shell Folders\\Fonts
```
Comment (QuLogic) [inline lib/matplotlib/font_manager.py]: The first one is supposed to be doubled, so it should be _4_ when escaped, but maybe it'd be simpler to make it a raw string instead?

File: lib/matplotlib/mathtext.py [hunk 4]
```diff
@@ -749,12 +749,12 @@ def _get_glyph(self, fontname, font_class, sym, fontsize, math=True):
         r'>'         : [('cal', 'i'), ('ex', 'E')]
         }
 
-    for alias, target in [('\leftparen', '('),
-                          ('\rightparent', ')'),
-                          ('\leftbrace', '{'),
-                          ('\rightbrace', '}'),
-                          ('\leftbracket', '['),
-                          ('\rightbracket', ']'),
+    for alias, target in [('\\leftparen', '('),
```
Comment (QuLogic) [inline lib/matplotlib/mathtext.py]: The last ones are raw strings; better to do that for consistency.

File: lib/matplotlib/patches.py [hunk 5]
```diff
@@ -1803,7 +1803,7 @@ def _pprint_styles(_styles):
     (stylename : styleclass), return a formatted string listing all the
     styles. Used to update the documentation.
     """
-    names, attrss, clss = [], [], []
+    ddnames, attrss, clss = [], [], []
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: Why?

File: lib/matplotlib/tests/test_subplots.py [hunk 6]
```diff
@@ -111,7 +111,7 @@ def test_exceptions():
     # the point of this test is to ensure that this raises.
     with warnings.catch_warnings():
         warnings.filterwarnings('ignore',
-                                message='.*sharex\ argument\ to\ subplots',
+                                message=r'.*sharex\ argument\ to\ subplots',
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_subplots.py]: What is `\ `? I don't think this escape even means anything in a regular expression.

File: lib/matplotlib/tests/test_text.py [hunk 7]
```diff
@@ -30,7 +30,7 @@ def find_matplotlib_font(**kw):
     from matplotlib.font_manager import FontProperties, findfont
     warnings.filterwarnings(
         'ignore',
-        ('findfont: Font family \[u?\'Foo\'\] not found. Falling back to .'),
+        r'findfont: Font family \[u?\'Foo\'\] not found. Falling back to .',
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_text.py]: I think the internal quotes no longer need to be escaped here now.

File: tests.py [hunk 8]
```diff
@@ -14,6 +14,27 @@
...
+    # The warnings need to be before any of matplotlib imports, but after
+    # setuptools (if present) which has syntax error with the warnings enabled.
+    # Filtering by module does not work as this will be raise by Python itself.
+    # so `module=matplotlib.*` is out of questions.
+
+    import warnings
+
+    # Python 3.6 deprecate invalid character-pairs \A, \* ... in non raw-strings
+    # and other things. Let's not re-introduce them
+    warnings.filterwarnings('error', category=DeprecationWarning)
```
Comment (QuLogic) [inline tests.py]: Though I prefer not to introduce any deprecations either, this seems a bit broad for this PR.

File: lib/matplotlib/tests/test_text.py (hunk 7 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_text.py]: Use double quotes instead?

Comment (QuLogic) [inline lib/matplotlib/tests/test_text.py]: Err, actually, I don't know what I was thinking here, but the internal quotes still need to be escaped unless using double quotes as @anntzer suggested.

File: lib/matplotlib/backends/backend_ps.py (hunk 1 above)
Comment (Carreau) [inline lib/matplotlib/backends/backend_ps.py]: Sorry, autopep8 script doing more than asked.

File: lib/matplotlib/font_manager.py (hunk 3 above)
Comment (Carreau) [inline lib/matplotlib/font_manager.py]: Oh.. ok, I did not remembered `\\` was a thing on windows.

File: lib/matplotlib/mathtext.py (hunk 4 above)
Comment (Carreau) [inline lib/matplotlib/mathtext.py]: sure. Done.

File: lib/matplotlib/patches.py (hunk 5 above)
Comment (Carreau) [inline lib/matplotlib/patches.py]: Hum.. vim, probably though I was typing in another window.

Comment (Carreau) [inline lib/matplotlib/patches.py]: though these 3 attributes are not use at all in the function...

File: tests.py [hunk 9]
```diff
@@ -14,6 +14,28 @@
...
 if __name__ == '__main__':
+
+    try:
+        import setuptools
+    except ImportError:
+        pass
+
+    # The warnings need to be before any of matplotlib imports, but after
+    # setuptools (if present) which has syntax error with the warnings enabled.
+    # Filtering by module does not work as this will be raise by Python itself.
```
Comment (anntzer) [inline tests.py]: typo: raised

Comment (Carreau) [inline tests.py]: Done.

**Comment by Carreau** [comment]
Rebased to remove conflicts.

**Comment by Carreau** [comment]
Rebased again and use backtick

File: tests.py [hunk 10]
```diff
@@ -14,6 +14,28 @@
...
+
+    # Python 3.6 deprecate invalid character-pairs \A, \* ... in non raw-strings
+    # and other things. Let's not re-introduce them
+    warnings.filterwarnings('error', '.*invalid escape sequence.*',
+        category=DeprecationWarning)
+    warnings.filterwarnings(
+        'default',
+        '.*inspect.getargspec\(\) is deprecated.*',
+        category=DeprecationWarning)
+
```
Comment (jkseppan) [inline tests.py]: This only gets executed when running tests.py, and I think most tests have been migrated to pytest now. Is there a way to make pytest execute these filterwarnings calls?

Comment (Carreau) [inline tests.py]: [Yes](https://github.com/fschulze/pytest-warnings). It's supposed to get folded into Pytest itself at some point.

File: lib/matplotlib/collections.py [hunk 11]
```diff
@@ -394,7 +394,7 @@ def set_hatch(self, hatch):
         *hatch* can be one of::
 
           /   - diagonal hatching
-          \   - back diagonal
+          \\   - back diagonal
```
Comment (tacaswell) [inline lib/matplotlib/collections.py]: This one worries me as it is something I would expect to be read either on the web or in the terminal.  Probably better to make this a `r'''` string?

Comment (tacaswell) [inline lib/matplotlib/collections.py]: and I may be wrong here, escaping always ties me in knots....

Comment (Carreau) [inline lib/matplotlib/collections.py]: Done. I can't rebuild matplotlib today apparently, so can't test the html versions.

Comment (Carreau) [inline lib/matplotlib/collections.py]: Actually I managed to build the docs, and the `\` in `Accept` below needed to be escaped more.
That should be fixed.

File: lib/matplotlib/collections.py [hunk 12]
```diff
@@ -415,7 +415,7 @@ def set_hatch(self, hatch):
         can only be specified for the collection as a whole, not separately
         for each member.
 
-        ACCEPTS: [ '/' | '\\\\' | '|' | '-' | '+' | 'x' | 'o' | 'O' | '.' | '*' ]
+        ACCEPTS: [ '/' | '\\' | '|' | '-' | '+' | 'x' | 'o' | 'O' | '.' | '*' ]
```
Comment (QuLogic) [inline lib/matplotlib/collections.py]: It appears that this line is under 79 characters and this file now passes PEP8, so it needs to be removed from `lib/matplotlib/tests/test_coding_standards.py` because the build is failing.

Comment (QuLogic) [inline lib/matplotlib/collections.py]: Also, since you're going to have to push an update, I cancelled the OSX build which has been waiting for a while because we've got a long backlog.

File: lib/matplotlib/tests/test_text.py (hunk 7 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_text.py]: @Carreau As mentioned just above, can you switch to double quotes and remove the backslashes in the inner single quotes here?

**Comment by NelleV** [comment]
Thanks @Carreau 
4 core dev validating your PR: I believe this is a new record :)

**Comment by Carreau** [comment]
Thanks !
